### PR TITLE
TPU gradient clipping. 

### PR DESCRIPTION
**Pull request by srush**
Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the contributor guideline?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?
- [ ] If you made a notable change (that affects users), did you update the CHANGELOG?
What does this PR do?
Switches gradient clipping to TPU friendly style.

PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

Did you have fun?
Make sure you had fun coding

### CODE REVIEW

File: pytorch_lightning/trainer/training_tricks.py [hunk 1]
```diff
@@ -19,9 +20,30 @@ def get_model(self):
...
             model = self.get_model()
-            torch.nn.utils.clip_grad_norm_(model.parameters(), self.gradient_clip_val)
+            parameters = model.parameters()
+            max_norm = self.gradient_clip_val
+            norm_type = 2
+            if isinstance(parameters, torch.Tensor):
+                parameters = [parameters]
+            parameters = list(filter(lambda p: p.grad is not None, parameters))
+            max_norm = float(max_norm)
+            norm_type = float(norm_type)
```
Comment (Borda) [inline pytorch_lightning/trainer/training_tricks.py]:
```suggestion
```

File: pytorch_lightning/trainer/training_tricks.py [hunk 2]
```diff
@@ -19,9 +20,30 @@ def get_model(self):
...
 
     def clip_gradients(self):
+        # this code is a modification of torch.nn.utils.clip_grad_norm_
+        # with TPU support based on https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md
         if self.gradient_clip_val > 0:
             model = self.get_model()
-            torch.nn.utils.clip_grad_norm_(model.parameters(), self.gradient_clip_val)
+            parameters = model.parameters()
+            max_norm = self.gradient_clip_val
+            norm_type = 2
```
Comment (Borda) [inline pytorch_lightning/trainer/training_tricks.py]:
```suggestion
            max_norm = float(self.gradient_clip_val)
            norm_type = float(2.0)
```

File: pytorch_lightning/trainer/training_tricks.py [hunk 3]
```diff
@@ -19,9 +20,30 @@ def get_model(self):
...
+            if norm_type == math.inf:
+                total_norm = max(p.grad.data.abs().max() for p in parameters)
+            else:
+                device = parameters[0].device
+                total_norm = torch.zeros([], device=device if parameters else None)
+                for p in parameters:
+                    param_norm = p.grad.data.norm(norm_type) ** norm_type
+                total_norm.add_(param_norm)
+                total_norm = (total_norm ** (1. / norm_type))
+            clip_coef = torch.tensor(max_norm, device=device) / (total_norm + 1e-6)
```
Comment (Borda) [inline pytorch_lightning/trainer/training_tricks.py]: this `1e-6` is kind of epsilon to avoid divide by 0, right?
what about `sys.float_info.epsilon`, see https://stackoverflow.com/questions/9528421/value-for-epsilon-in-python

Comment (srush) [reply pytorch_lightning/trainer/training_tricks.py]: Is it okay if I keep this? It comes directly from the pytorch code which I am trying to stay close to.

Comment (Borda) [reply pytorch_lightning/trainer/training_tricks.py]: it is fine, I would just declare it in the module top like `EPSILON = 1e-6` so it is clear

Comment (Borda) [reply pytorch_lightning/trainer/training_tricks.py]: it is fine, I would just declare it in the module top like `EPSILON = 1e-6` so it is clear

Comment (Borda) [reply pytorch_lightning/trainer/training_tricks.py]: it is fine, I would just declare it in the module top like `EPSILON = 1e-6` so it is clear

Comment (williamFalcon) [reply pytorch_lightning/trainer/training_tricks.py]: isn’t 1e-6 zero when on 16 bit? can we make it 1e-5 when 16 bit is on?

Comment (srush) [reply pytorch_lightning/trainer/training_tricks.py]: Changed to 1e-6. Not sure what you are supposed to do with 16 bit (pytorch doesn't change the call)

Comment (Borda) [reply pytorch_lightning/trainer/training_tricks.py]: there is an option to check if it runs fp16 training `use_amp=hparams.use_16bit`

Comment (srush) [reply pytorch_lightning/trainer/training_tricks.py]: Should I do this or just make eps 1e-5?

Comment (srush) [reply pytorch_lightning/trainer/training_tricks.py]: Okay I changed it to check self.precisions

File: pytorch_lightning/trainer/training_tricks.py [hunk 4]
```diff
@@ -19,9 +23,31 @@ def get_model(self):
...
+                device = parameters[0].device
+                total_norm = torch.zeros([], device=device if parameters else None)
+                for p in parameters:
+                    param_norm = p.grad.data.norm(norm_type) ** norm_type
+                total_norm.add_(param_norm)
+                total_norm = (total_norm ** (1. / norm_type))
+            if self.precision == 16:
+                clip_coef = torch.tensor(max_norm, device=device) / (total_norm + EPSILON_FP16)
+            else:
+                clip_coef = torch.tensor(max_norm, device=device) / (total_norm + EPSILON)
```
Comment (Borda) [inline pytorch_lightning/trainer/training_tricks.py]:
```suggestion
            eps = EPSILON_FP16 if self.precision == 16 else EPSILON
            clip_coef = torch.tensor(max_norm, device=device) / (total_norm + eps)
```
